Object type tag 2569

knirirr opened the pull request:
The object_type server code has been deployed in dev.

knirirr committed:
Object type can be added and viewed. Tests run but incomplete. #2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Fixed coverage.

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Removed debugging info.

knirirr committed:
Object type available in basic search.

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Fixed tests.

knirirr committed:
Add "not found" object type when creating record. #2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Added object type to adv. search (no tests yet). #2569

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Added tests #2569.

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Removed console.log.

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Yet more console.log. #2569

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Redirect object not found doi to gitbook. #2569

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Update docs.

knirirr committed:
Added missing file.

knirirr committed:
Merge branch 'dev' into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

allysonlister commented:
![image](https://github.com/user-attachments/assets/0fe1f7c1-029b-4a90-9a05-deed7e167def)
I cannot see the new object tags at all. I note that the domain tags that match the object tags (see in the screenshot the maroon 'document' and 'publication' are visible, but that is to be expected as we haven't removed them from DRAO yet.

Shouldn't the object tags be visible in this screenshot? Again any record will do, but I'm in https://deploy-preview-2570--fairsharing.netlify.app/1561/edit

allysonlister commented:
I have randomly tried to save https://deploy-preview-2570--fairsharing.netlify.app/1561 without changing anything, to see if the "missing object type" appears in this edit or later on in the email that will be sent relating to minimal curation reqs.

However, the 'save and exit' i tried is hanging and just spinning:
![Screenshot From 2025-05-21 11-48-09](https://github.com/user-attachments/assets/cea20da1-8072-44e8-9cd9-9aed16f4224f)

knirirr committed:
Don't assign object_type on creation #2569

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Autogenerated documentation files

allysonlister commented:
The functionality of the edit interface works great, and all object types are present.

Please note the following:

1. When hovering over an object type pill, please can we have the IRI displayed as with the other tag types?
2. At some point in the future (perhaps via another ticket), @allysonlister should provide definitions to go into the edit interface, and also the hoverover text with the IR

Otherwise I think it's ready. If you are happy with some overlaps in terms for a couple of weeks (ie when you deploy the DRAO and SRAO tags will still be there until I can provide a new release), then I'm happy for this to be merged.

Note the specific minimal curation reqs etc that go with this object type, as defined in the document. thanks!

knirirr committed:
Tooltips for object type on record pages #2569

knirirr committed:
Autogenerated documentation files

knirirr commented:
>    When hovering over an object type pill, please can we have the IRI displayed as with the other tag types?

Yes; I have added those.

> Note the specific minimal curation reqs etc that go with this object type, as defined in the document. thanks!

That's at https://github.com/FAIRsharing/FAIRsharing-API/blob/object_type_as_tag/app/models/concerns/completable.rb#L42-L55
I've now added the text from the Google doc.

It would perhaps be best to wait until this is merged until @allysonlister is back. But, at that point, it should be easy enough to do if there are no further changes (for now).

knirirr committed:
Updated docs.

knirirr committed:
Autogenerated documentation files

knirirr committed:
Merge branch 'dev' into object_type_tag_2569

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr committed:
Updated docs.

knirirr committed:
Autogenerated documentation files

knirirr committed:
Typo fix. #2569

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

allysonlister commented:
Perhaps as a separate ticket (to get this merged) you could create an issue to add the object type pills to the search results, both grid and stack views? thanks!

knirirr commented:
I might as well do that here; I'll need to wait for approval on the PRs and Ramon won't be back until tomorrow.

allysonlister commented:
When adding an object type, the hover-over text exists for all other subject tags (either newly-added or already there) AND for already-present object types. However, the hover-over does NOT exist for newly-added object types:

You cannot see my mouse, but it is hovering over 'software source code' here:
![image](https://github.com/user-attachments/assets/6e594fdd-732b-4ab1-a7b4-a699caf201bc)

Note that the hoverover works for already-present object types
![Screenshot From 2025-06-02 10-51-55](https://github.com/user-attachments/assets/70e3cb4a-07bb-41b2-8b8c-f64757fe7151)

And the hover-over works for already-present as well as newly-added tags of other types, e.g. this domain here:
![Screenshot From 2025-06-02 10-51-51](https://github.com/user-attachments/assets/69df7470-c94a-47c4-a156-285c8bc2cb33)

allysonlister commented:
> Add definitions to the object types as per the definition (NOT the example or other info) in https://fairsharing.gitbook.io/fairsharing/record-sections-and-fields/general-information/object-types

These differ from the types I have in the application. Are the object types at that list the correct ones, or are there further changes planed? If they are correct I can edit what I have and re-implement them (including definitions).

BTW, I will have to code them so they have the ability to have multiple definitions, so the code which already exists can handle them, even though in practice they should only have one.

knirirr committed:
Added object type tags to search results. #2569

knirirr committed:
Autogenerated documentation files

knirirr committed:
Removed console.log.

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

allysonlister commented:
The name labels in gitbook have different mixed casing, but they should all exactly match those listed in https://github.com/OSTrails/digital-object-commons/blob/main/object_types_draft.csv. Other than minor differences in the labels, do you see anything major here?

Which ones have multiple definitions? Not sure why you'd need to code that?

Sorry I didn't reply sooner - you edited my comment rather than making a new one as far as I can tell, so I didn't get a notification?

allysonlister commented:
let me know when I can check again :)

knirirr commented:
I'd prefer to edit rather than comment again as these threads can get a little long.
Some other tags (Subject or Domain) can have multiple definitions so anything to do with tooltips on tags will be expecting that and ObjectType will have to support it as well. 
Anyway, I am attempting to proceed with adding these definitions.

knirirr committed:
Get definitions of object_types from server. #2569.

allysonlister commented:
that's fantastic, thank you

knirirr committed:
Tooltips with definitions for object_type #2569

knirirr committed:
Modified query.

knirirr commented:
This latest feature (definitions) required some changes to the server code and also reindexing of ObjectTypes. However, everything above is working on the Netlify build, as far as I can tell. Please let me know if there's still anything unexpected.
If it's all in order and there's nothing more to add then this could be deployed tomorrow.

allysonlister commented:
Thank you this is looking great! 

Here as some very small tweaks for the definitions, as I note that the definitions in gitbook refer to object types "above" another in the webpage.

Please can you change the following definitions?
* document: A generic object type for any electronic file or document that does not fit into any of the other document sub-types provided (publication, questionnaire/survey or data management plan). This object type includes any kind of document or text file, as well as training material e.g. presentations, and coursework.
* image: Where the digital research object is (i) image based, e.g. microscope, telescope, gel images, or (ii) highly visual, e.g. maps. Please note "multimedia" (for video or audio) is a separate tag in FAIRsharing. If the object type is a digital representation (including images, 3d scans) of a physical object (e.g. a natural history collection), then please use the "physical object/material" tag instead.
* Other object type: This type should only be used if absolutely none of the other types are suitable and the resource is not agnostic of any particular object type. In other words, only use this when the object type for the resource is not among any of the other types provided.
* Object type not found: Only used for certain older records, when the FAIRsharing team can no longer find information on the object type (e.g. older deprecated records).

allysonlister commented:
I am happy for this to be deployed once you and @ramGranell are ready, e.g. tomorrow. thanks!

knirirr commented:
> Please can you change the following definitions?

Yes, but as they're objects in the database the easiest way to do that at this point will be to delete and re-create them.
Once they're deployed in production then a rake task would be needed, so with luck they will remain stable. 
I will deal with that this morning.

ramGranell commented:
I have found the following error when I performed these steps:
I edited record 1551 adding three object types when there was none
I save them correctly, I edited again the same record removing the three object types and I try to save the record. It was not able to save it having the same error:

TypeError: this.currentRecord is undefined
    saveRecord GeneralInformation.vue:270
    checkTypeChange GeneralInformation.vue:234
    click GeneralInformation.vue:1
    VueJS 4
    click VBtn.ts:163
    VueJS 3

You can replicate this going to the record https://deploy-preview-2570--fairsharing.netlify.app/1551/edit remove the three object types and try to save it. It also happens when I created a new record.

I create a new record, a database, I go to edit it, I added an object type and I cannot save as it complains about taxonomy. Is taxonomy the only minimal curation record field that it is mandatory? Why is only complaining about taxonomy? Maybe this is not related to this ticket, if so I can open a new ticket.

Other thing, in the question mark of the "Object type" appears "The curators will decide what text will be here." Can we decide what to write before approving it?

The rest of the behaviour seems correctly, but I noted two things: 1) Object type does not appear in the normal search (https://deploy-preview-2570--fairsharing.netlify.app/search ) but I don't know if we wanted to have it. It seems logical to have it as we have subject, domain and species and 2) Object type does not appear in the advance search option and it seems to have as commented before.

knirirr committed:
Removed now unneeded object type presence check. #2569

knirirr committed:
Autogenerated documentation files

knirirr commented:
> It was not able to save it having the same error:

This was some code relating to enforcing the addition of object types, which I have now removed as these will be dealt with by the email.

> Why is only complaining about taxonomy? Maybe this is not related to this ticket, if so I can open a new ticket.

I can't find the relevant ticket(*) but my recollection is that a taxonomy field is always required and that it should be "not applicable" (or whatever the relevant tag is) if not relevant. Suggestions to make it optional have been rejected on the grounds that it would then not be possible to tell if taxonomy had simply been omitted by an editor or whether it was not relevant. 

> Other thing, in the question mark of the "Object type" appears "The curators will decide what text will be here." Can we decide what to write before approving it?

Yes, please do. There might be some existing text somewhere but I'm not sure what/where.  

> Object type does not appear in the normal search 
> Object type does not appear in the advance search option 

If you want any of these to be done please decide whether it will be as part of this PR or separate tickets to be dealt with once this one is merged.


(*) The discussion might have been in a PR, which would make it harder to find.

ramGranell commented:
About taxonomy is a mandatory field for the all the records as it was established in the minimal curation fields. Then, record maintainers/creators would receive and email indicating that the field is missing in the case they have not added it. I don't think a server error is needed as it is inconsistent with the rest of the mandatory minimal curation fields such as subjects, that it does not have a server error. I think that with the minimal curation field system, this server error can be removed, do you agree @allysonlister ?

knirirr commented:
> I think that with the minimal curation field system, this server error can be removed, do you agree @allysonlister ?

@allysonlister if you want to do this, please would you or @ramGranell create a separate (server) ticket for it. I could then proceed with completing this (various other items from https://github.com/FAIRsharing/fairsharing.github.io/pull/2570#issuecomment-2934090775).

allysonlister commented:
> > Other thing, in the question mark of the "Object type" appears "The curators will decide what text will be here." Can we decide what to write before approving it?
> 
> Yes, please do. There might be some existing text somewhere but I'm not sure what/where.

The text is here: https://docs.google.com/document/d/1pOtArzTjMx4bIuVIDQgPJFttoQMd85oi3oifYY19T_M/edit?tab=t.nphrwqf21val, although it does only mention it is for missing fields, it can definitely be used for the ? text. Here's a shorter version appropriate for the ?: "Object types define the type of digital research object that is in scope for the resource. At least one object type must be provided for all record types other than collections."

allysonlister commented:
> > Object type does not appear in the normal search
> > Object type does not appear in the advance search option
> 
> If you want any of these to be done please decide whether it will be as part of this PR or separate tickets to be dealt with once this one is merged.

Yes we definitely want to have object types in advanced search and in the regular search facets. Happy for it to be a separate ticket so we can get this work out, but consider it at least medium priority. thanks!

allysonlister commented:
> About taxonomy is a mandatory field for the all the records as it was established in the minimal curation fields. Then, record maintainers/creators would receive and email indicating that the field is missing in the case they have not added it. I don't think a server error is needed as it is inconsistent with the rest of the mandatory minimal curation fields such as subjects, that it does not have a server error. I think that with the minimal curation field system, this server error can be removed, do you agree @allysonlister ?

I hesitate to change it as there are other things that are server errors (I believe zero subject tags is also a server error?) so rather than doing it piecemeal we should either leave it or re-evaluate everything that is currently a server error. either way it would be a separate ticket.

knirirr committed:
Changed some tooltip text.

knirirr committed:
Merge branch 'object_type_tag_2569' of github.com:FAIRsharing/fairsharing.github.io into object_type_tag_2569

knirirr commented:
I have changed that text. 
Please click the approval button(s) here and at https://github.com/FAIRsharing/FAIRsharing-API/pull/1245 if you are now satisfied - thanks.

ramGranell commented:
> > About taxonomy is a mandatory field for the all the records as it was established in the minimal curation fields. Then, record maintainers/creators would receive and email indicating that the field is missing in the case they have not added it. I don't think a server error is needed as it is inconsistent with the rest of the mandatory minimal curation fields such as subjects, that it does not have a server error. I think that with the minimal curation field system, this server error can be removed, do you agree @allysonlister ?
> 
> I hesitate to change it as there are other things that are server errors (I believe zero subject tags is also a server error?) so rather than doing it piecemeal we should either leave it or re-evaluate everything that is currently a server error. either way it would be a separate ticket.

The only thing that causes server error when there is not a value after a record is created (a record to be created needs to have a name, a homepage, registry and type and a description with a minimal number of characters) is taxonomies. The rest of mandatory fields described in the minimal curation list of fields do not provoke a sever error when they do not exist. They are notified in the email as incomplete fields. I consider that this is an inconsistent behaviour with the rest of mandatory fields such as subject or object type.

knirirr commented:
>  I consider that this is an inconsistent behaviour with the rest of mandatory fields such as subject or object type.

Please open a separate (server) ticket for this, and once there is a decision I'll queue this up to deal with after the existing tickets.

allysonlister commented:
> > I consider that this is an inconsistent behaviour with the rest of mandatory fields such as subject or object type.
> 
> Please open a separate (server) ticket for this, and once there is a decision I'll queue this up to deal with after the existing tickets.

Thanks. @ramGranell please could you create a discussion ticket for this, and a second ticket for object types in advanced search and regular search facets?

allysonlister reviewed (approved):
thank you - all set!
